File: apps/yryvu-app/src/ipc/integrations/clone.ts
// SPDX-License-Identifier: AGPL-3.0-or-later

import { invoke } from "@tauri-apps/api/core";

/// Personal account vs organization / group ownership of a clone
/// candidate. Drives the dropdown's personal-first ordering.
export type OwnerKind = "user" | "organization";

/// Cross-provider repo candidate surfaced by the Clone dialog's
/// per-provider sub-tabs (#374). Backend mirrors the same shape.
export interface CloneRepoCandidate {
  owner: string;
  ownerKind: OwnerKind;
  name: string;
  fullName: string;
  cloneUrlHttps: string;
  cloneUrlSsh?: string;
  isPrivate: boolean;
  description?: string;
  defaultBranch?: string;
}

/// List every repo the authenticated user can clone via
/// `integrationType`. Bitbucket / Azure return `NotImplemented` from
/// the backend; the frontend renders a "backend pending" hint instead
/// of attempting the call.
export function integrationListCloneCandidates(
  integrationType: string,
): Promise<CloneRepoCandidate[]> {
  return invoke<CloneRepoCandidate[]>("integration_list_clone_candidates", {
    integrationType,
  });
}

File: crates/yryvu-bridge/src/commands/integrations/clone.rs
// SPDX-License-Identifier: AGPL-3.0-or-later

use tauri::AppHandle;

use crate::integrations::{self, CloneRepoCandidate};

use super::{host_for, load_auth_and_host};

/// List every repo the authenticated user can clone via
/// `integration_type`. Powers the Clone dialog's per-provider sub-tab
/// (#374). Returns a flat list — frontend groups by `owner` for the
/// GK-style org-headered dropdown.
#[tauri::command]
pub async fn integration_list_clone_candidates(
    app: AppHandle,
    profile_id: Option<String>,
    integration_type: String,
) -> Result<Vec<CloneRepoCandidate>, String> {
    let auth = load_auth_and_host(&app, profile_id.as_deref(), &integration_type).await?;
    let hostname = host_for(&integration_type, &auth);
    integrations::list_clone_candidates(&integration_type, &auth.token, hostname)
        .await
        .map_err(|e| e.to_string())
}
